sklearn/neighbors/lof.py

```diff
@@ -0,0 +1,313 @@
+        a sample and those of its k-nearest neighbors.
+
+        Parameters
+        ----------
+
+        X : array-like, shape (n_samples, n_features), optional
+            The query sample or samples to compute the Local Outlier Factor
+            w.r.t. the training samples. If None, makes prediction on the
+            training data without considering them as their own neighbors.
+
+        Returns
+        -------
+        lof_scores : array of shape (n_samples,)
+            The Local Outlier Factor of each input samples. The lower,
+            the more abnormal.
+        """
+        check_is_fitted(self, ["threshold_", "decision_function_train_",
+                               "_k_distance_value_fit_X_",
+                               "neighbors_indices_fit_X_"])
+
+        # if X is None, make prediction on X=X_train without considered each
+        # query point as its own neighbor.
+        if X is None:
+            return self.decision_function_train_
+
+        else:
+            X = check_array(X, accept_sparse='csr')
+            return -self._local_outlier_factor(X)  # as bigger is better
+
+    def _k_distance(self, X=None):
+        """
+        Compute the k_distance and the neighborhood of query samples wrt
+        training samples self._fit_X.
+        If X=None, neighbors of each sample self._fit_X are returned.
+        In this case, the query point is not considered its own neighbor.
+        """
+        distances, neighbors_indices = self.kneighbors(
+            X=X, n_neighbors=self.n_neighbors)
+        k_dist = distances[:, self.n_neighbors - 1]
+
+        return k_dist, neighbors_indices
+
+    def _local_reachability_density(self, X, neighbors_indices):
+        """The local reachability density (LRD)
+
+        The LRD of a sample is the inverse of the average reachability
+        distance of its k-nearest neighbors.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The samples to compute the LRD w.r.t. self._fit_X
+            If None, compute the LRD of self._fit_X w.r.t. itself.
+            (In this case samples are not considered in their own
+            neighborhood)
+
+        neighbors_indices : array-like of shape (n_samples, self.n_neighbors)
+            neighbors indices of X (or of self._fit_X if X is None)
+            Added as a parameter to avoid re-computing it with
+            self._k_distance(X)[1] when X is not None (if X is None, it is
+            equal to self.neighbors_indices_fit_X_)
+
+        Returns
+        -------
+        local_reachability_density : array, shape (n_samples,)
+            The local reachability density of each sample.
+        """
+        if X is None:
+            X = self._fit_X
+
+        n_samples = X.shape[0]
+        reach_dist_array = np.zeros((n_samples, self.n_neighbors))
+
+        for j in range(n_samples):
```

Comment:
It seems that we can. I tried the following code

```
    def _local_reachability_density(self, X, neighbors_indices):

        dist, ind = self.kneighbors(X)
        dist_k = self._k_distance_value_fit_X_[ind]

        reach_dist_array = np.maximum(dist, dist_k)

        return 1. / (np.mean(reach_dist_array, axis=1) + 1e-10)
```

All tests are OK, running the benchmark is faster for all datasets except SA (because of the `self.kneighbors(X)`? if it is a very large dataset?). And when I run my script using anomaly detection algorithms everything works the same and it is clearly faster.

With for loop version:
![bench_lof_1](https://cloud.githubusercontent.com/assets/15966638/15929018/ac68daaa-2e4a-11e6-89c1-a8b5d8a89f35.png)

With vectorized version:
![bench_lof_2](https://cloud.githubusercontent.com/assets/15966638/15929031/c63dd26e-2e4a-11e6-91a7-8a955ca679a7.png)

Note that I cannot reproduce your ROC curves for SF and HTTP when running bench_lof.py (with both versions). I don't know why...